### PR TITLE
feature/320_die_if_no_logs

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue https://github.com/telefonicaid/fiware-connectors/issues/320
* Fixes issue https://github.com/telefonicaid/fiware-connectors/issues/302
* 100% unit tests passed
```
Results :
Tests run: 45, Failures: 0, Errors: 0, Skipped: 0
```
* (Unofficial) e2e tests passed (emulating a persistence error, ttl=2)
```
Starting an ordered shutdown of Cygnus
Stopping sources
Stopping http-source (lyfecycle state=START)
There are 1 events within hdfs-channel, Cygnus cannnot shutdown yet
Waiting 5 seconds
There are 1 events within hdfs-channel, Cygnus cannnot shutdown yet
Waiting 5 seconds
There are 1 events within hdfs-channel, Cygnus cannnot shutdown yet
Waiting 5 seconds
All the channels are empty
Stopping channels
Stopping hdfs-channel (lyfecycle state=START)
Stopping sinks
Stopping hdfs-sink (lyfecycle state=START)
```
* Ths PR must be twined into develop
* Assignee: @fgalan , but LGTM from @gtorodelvalle is also required

### CODE REVIEW

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java [hunk 1]
```diff
@@ -40,7 +48,14 @@ public CygnusLogger(String name) {
      * @param message
      */
     public void info(String componentName, String message) {
-        info("[" + componentName + "] " + message);
+        try {
+            info("[" + componentName + "] " + message);
+        } catch (Exception e) {
+            if (shutdown) {
+                System.err.println("A problem with the logging system was found... shutting down Cygnus right now!");
```
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Maybe you should also print the Exception trace/information to provide more information to the administrator reading the mesage?

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java [hunk 2]
```diff
@@ -49,7 +64,14 @@ public void info(String componentName, String message) {
      * @param message
      */
     public void error(String componentName, String message) {
-        error("[" + componentName + "] " + message);
+        try {
+            error("[" + componentName + "] " + message);
+        } catch (Exception e) {
+            if (shutdown) {
+                System.err.println("A problem with the logging system was found... shutting down Cygnus right now!");
```
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Maybe you should also print the Exception trace/information to provide more information to the administrator reading the mesage?

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java [hunk 3]
```diff
@@ -58,7 +80,14 @@ public void error(String componentName, String message) {
      * @param message
      */
     public void fatal(String componentName, String message) {
-        fatal("[" + componentName + "] " + message);
+        try {
+            fatal("[" + componentName + "] " + message);
+        } catch (Exception e) {
+            if (shutdown) {
+                System.err.println("A problem with the logging system was found... shutting down Cygnus right now!");
```
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Maybe you should also print the Exception trace/information to provide more information to the administrator reading the mesage?

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java [hunk 4]
```diff
@@ -67,7 +96,14 @@ public void fatal(String componentName, String message) {
      * @param message
      */
     public void warn(String componentName, String message) {
-        warn("[" + componentName + "] " + message);
+        try {
+            warn("[" + componentName + "] " + message);
+        } catch (Exception e) {
+            if (shutdown) {
+                System.err.println("A problem with the logging system was found... shutting down Cygnus right now!");
```
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Maybe you should also print the Exception trace/information to provide more information to the administrator reading the mesage?

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java [hunk 5]
```diff
@@ -76,7 +112,14 @@ public void warn(String componentName, String message) {
      * @param message
      */
     public void debug(String componentName, String message) {
-        debug("[" + componentName + "] " + message);
+        try {
+            debug("[" + componentName + "] " + message);
+        } catch (Exception e) {
+            if (shutdown) {
+                System.err.println("A problem with the logging system was found... shutting down Cygnus right now!");
```
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Maybe you should also print the Exception trace/information to provide more information to the administrator reading the mesage?

(BTW, I have copy-paste the same comment 5 times... maybe the code could be refactored to have the print & exit logic in one place :)

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/nodes/CygnusApplication.java [hunk 6]
```diff
@@ -124,14 +166,14 @@ public static void main(String[] args) {
             boolean reload = !commandLine.hasOption("no-reload-conf");
 
             if (commandLine.hasOption('h')) {
-                new HelpFormatter().printHelp("flume-ng agent", options, true);
+                new HelpFormatter().printHelp("cygnus-flume-ng agent", options, true);
```
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/nodes/CygnusApplication.java]: This is in sync with the documentation?

Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/nodes/CygnusApplication.java]: Yes, simply this help line was not updated. But the README explains that we have a `cygnus-flume-ng` specific script.

**Comment by gtorodelvalle** [comment]
Sorry for commenting on the commits :( I left-click-opened them and start commenting there :( I will comment on the pull request "File changed" tab from now on ;)

Anyhow, I think all my comments are included so LGTM on my side ;) Thanks @frbattid ;)

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/nodes/CygnusApplication.java (hunk 6 above)
Comment (fgalan) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/nodes/CygnusApplication.java]: NTC

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java (hunk 1 above)
Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Fixed in https://github.com/telefonicaid/fiware-connectors/commit/083ba577850ad21c798abe1c183f794ee2f2c4ea

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java (hunk 2 above)
Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Fixed in https://github.com/telefonicaid/fiware-connectors/commit/083ba577850ad21c798abe1c183f794ee2f2c4ea

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java (hunk 3 above)
Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Fixed in https://github.com/telefonicaid/fiware-connectors/commit/083ba577850ad21c798abe1c183f794ee2f2c4ea

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java (hunk 4 above)
Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Fixed in https://github.com/telefonicaid/fiware-connectors/commit/083ba577850ad21c798abe1c183f794ee2f2c4ea

File: flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java (hunk 5 above)
Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: Fixed in https://github.com/telefonicaid/fiware-connectors/commit/083ba577850ad21c798abe1c183f794ee2f2c4ea

Comment (frbattid) [inline flume/src/main/java/es/tid/fiware/fiwareconnectors/cygnus/log/CygnusLogger.java]: The `traceAndExit()` function was fixed in https://github.com/telefonicaid/fiware-connectors/commit/0893348c3335775f7d995af54c3394081db9e2b4

**Comment by frbattid** [comment]
Please ignore the following commits:

* MODIFY CygnusLogger, now slf4j with bindings to log4j12 is used https://github.com/telefonicaid/fiware-connectors/commit/1ebd01f8f1eccba3d0fa4f171a616e7fb2e05eaa
* ADD slf4j-api and slf4j-log4h12 dependencies to the pom.xml https://github.com/telefonicaid/fiware-connectors/commit/a9484b86572b2377914f54e9888e29a8e9ff4f3a
* MODIFY all the classes for using CygnusLogger instead of Logger https://github.com/telefonicaid/fiware-connectors/commit/c82799cb9fd499ec8dded4762f421a742ce50495

**Comment by fgalan** [comment]
LGTM

(Although @gtorodelvalle also told LGTM, not sure if I should merge, given the 8 commits since that point. If you want me to merge, just tell me).

**Comment by gtorodelvalle** [comment]
LGTM on my side too after @frbattid comments to mine ;)

Passing the merging over to @fgalan in case you want to do a final check with regard to those new commits :) Thanks!

**Comment by gtorodelvalle** [comment]
After discussing it with @frbattid , we're pretty confident about the PR and I am merging it ;) Thanks @frbattid ! :)

**Pull request merged**